dsgrid/registry/registry_database.py

```diff
@@ -0,0 +1,416 @@
+
+    def url(self):
+        return f"http://{self.hostname}:{self.port}"
+
+
+class RegistryDatabase:
+    """Database containing a dsgrid registry"""
+
+    def __init__(self, client):
+        self._client = client
+        self._graph = client.graph(GRAPH)
+
+    @classmethod
+    def create(cls, conn: DatabaseConnection, data_path: Path):
+        """Create a database to store a dsgrid registry."""
+        client = ArangoClient(hosts=f"http://{conn.hostname}:{conn.port}")
+        sys_db = client.db("_system", username=conn.username, password=conn.password)
+        if not sys_db.create_database(conn.database):
+            raise Exception(f"Failed to create the database {conn.database}")
+        client = client.db(conn.database, username=conn.username, password=conn.password)
+        client.create_collection("data_path")
+        # For AWS there will be other fields.
+        client.collection("data_path").insert({"data_path": str(data_path)})
+        data_path.mkdir(exist_ok=True)
+        registry_file = data_path / "registry.json5"
+        registry_file.write_text(conn.json(indent=2))
+
+        graph = client.create_graph("registry")
+        graph.create_vertex_collection(Collection.PROJECT_ROOTS.value)
+        graph.create_vertex_collection(Collection.DATASET_ROOTS.value)
+        graph.create_vertex_collection(Collection.DIMENSION_ROOTS.value)
+        graph.create_vertex_collection(Collection.DIMENSION_MAPPING_ROOTS.value)
+        graph.create_vertex_collection(Collection.PROJECTS.value)
+        graph.create_vertex_collection(Collection.DATASETS.value)
+        graph.create_vertex_collection(Collection.DIMENSIONS.value)
+        graph.create_vertex_collection(Collection.DIMENSION_MAPPINGS.value)
+        dimension_types = graph.create_vertex_collection(Collection.DIMENSION_TYPES.value)
+        for dim_type in DimensionType:
+            dimension_types.insert({"_key": dim_type.value})
+
+        graph.create_edge_definition(
+            edge_collection=Edge.CONTAINS.value,
+            from_vertex_collections=[
+                Collection.PROJECTS.value,
+                Collection.DATASETS.value,
+                Collection.DIMENSION_MAPPINGS.value,
+            ],
+            to_vertex_collections=[
+                Collection.DIMENSIONS.value,
+                Collection.DIMENSION_MAPPINGS.value,
+                Collection.DATASETS.value,
+            ],
+        )
+        graph.create_edge_definition(
+            edge_collection=Edge.DERIVES.value,
+            from_vertex_collections=[Collection.DATASETS.value],
+            to_vertex_collections=[Collection.DATASETS.value],
+        )
+        graph.create_edge_definition(
+            edge_collection=Edge.UPDATED_TO.value,
+            from_vertex_collections=[
+                Collection.DIMENSION_ROOTS.value,
+                Collection.DIMENSIONS.value,
+            ],
+            to_vertex_collections=[Collection.DIMENSION_ROOTS.value],
+        )
```

Comment:
Other things can be updated as well, right? So they should also be in the UPDATED_TO edge collection? But that again brings up support for specific pairs rather than any of the from's to any of the to's ...

Reply:
I don't really know how this works. This library makes you specify at least one from/to pair but doesn't seem to do any checking. I left a TODO and will revisit it later.